File: tinkerpop3/src/test/java/io/shiftleft/overflowdb/testdomains/gratefuldead/Artist.java
package io.shiftleft.overflowdb.testdomains.gratefuldead;

import io.shiftleft.overflowdb.NodeFactory;
import io.shiftleft.overflowdb.NodeRef;
import io.shiftleft.overflowdb.OdbGraph;

public class Artist extends NodeRef<ArtistDb> {
  public static final String label = "artist";
  public static final String NAME = "name";

  public Artist(OdbGraph graph, long id) {
    super(graph, id);
  }

  public String name() {
    return get().name();
  }

  @Override
  public String label() {
    return Artist.label;
  }

  public static NodeFactory<ArtistDb> factory = new NodeFactory<ArtistDb>() {
    @Override
    public String forLabel() {
      return Artist.label;
    }

    @Override
    public ArtistDb createNode(NodeRef<ArtistDb> ref) {
      return new ArtistDb(ref);
    }

    @Override
    public Artist createNodeRef(OdbGraph graph, long id) {
      return new Artist(graph, id);
    }
  };
}

File: traversals/src/test/scala/io/shiftleft/overflowdb/traversals/testdomains/gratefuldead/Artist.scala
package io.shiftleft.overflowdb.traversals.testdomains.gratefuldead

import io.shiftleft.overflowdb.traversals.{NodeRefOps, Traversal}
import io.shiftleft.overflowdb.{NodeFactory, NodeLayoutInformation, NodeRef, OdbGraph}

import scala.jdk.CollectionConverters._

class Artist(graph: OdbGraph, id: Long) extends NodeRef[ArtistDb](graph, id) with NodeRefOps[Artist] {
  override def label: String = Artist.Label

  def name: String = get.name

  /* Artist <-- sungBy --- Song */
  def sangSongs: Traversal[Song] = get.sangSongs
}

object Artist {
  val Label = "artist"

  object Properties {
    val Name = "name"
    val all: Set[String] = Set(Name)
    val allAsJava: java.util.Set[String] = all.asJava
  }

  val factory: NodeFactory[ArtistDb] = new NodeFactory[ArtistDb]() {
    override def forLabel: String = Artist.Label
    override def createNode(ref: NodeRef[ArtistDb]) = new ArtistDb(ref)
    override def createNodeRef(graph: OdbGraph, id: Long) = new Artist(graph, id)
  }

  val layoutInformation: NodeLayoutInformation =
    new NodeLayoutInformation(
      Properties.allAsJava,
      Nil.asJava,
      List(SungBy.layoutInformation, WrittenBy.layoutInformation).asJava
    )
}



